router/src/main.c

```diff
@@ -39,18 +39,26 @@ static qd_dispatch_t *dispatch = 0;
 static qd_log_source_t *log_source = 0;
 static const char* argv0 = 0;
 
+/**
+ * Configures the handler function. Specify SIG_IGN to ignore incoming signals.
+ */
+static void install_signal_handler(void (*handler)(int))
+{
+    signal(SIGHUP, handler);
+    signal(SIGQUIT, handler);
+    signal(SIGTERM, handler);
+    signal(SIGINT, handler);
+}
 
 /**
  * This is the OS signal handler, invoked on an undetermined thread at a completely
  * arbitrary point of time.
```

Comment:
Not necessarily. It seems to be possible, with pthreads, to ensure that there is only one thread available for handling signals. The OS will then be forced to deliver signals to it.

Alternative solution is this: the signal handler submits an action to some queue, and the action queue consumer then takes care of handling the signal.